tests/api/test_account_events.py

```diff
@@ -0,0 +1,223 @@
+from dataclasses import dataclass
+from typing import Optional
+
+import graphene
+import pytest
+from django.db import models
+
+from saleor.account import events as account_events
+
+from .utils import get_graphql_content
+
+QUERY_CUSTOMER_EVENTS = """
+query customerEvents($customerId: ID!) {
+  user(id: $customerId) {
+    id
+    events {
+      id
+      type
+      user {
+        id
+      }
+      message
+      count
+      order {
+        id
+      }
+      orderLine {
+        id
+      }
+    }
+  }
+}
+"""
+
+
+def test_account_events_are_properly_restricted(
+    staff_api_client, permission_manage_users, staff_user
+):
+    """Ensure that retrieve events of a customer is properly restricted."""
+
+    # Retrieve the staff user's id
+    customer_id = graphene.Node.to_global_id("User", staff_user.id)
+
+    # Test both with and without proper permissions
+    content = get_graphql_content(
+        staff_api_client.post_graphql(
+            QUERY_CUSTOMER_EVENTS,
+            variables={"customerId": customer_id},
+            permissions=[permission_manage_users],
+        )
+    )["data"]["user"]
+
+    # Ensure the correct data is returned when the request is valid
+    assert content["id"] == customer_id
+    assert content["events"] == []
+
+
+@dataclass
+class _ExpectedCustomerEventData:
+    id: str
+    user: str
+    count: Optional[int]
+    message: Optional[str]
+    order: Optional[str]
+    orderLine: Optional[str]  # noqa
+    type: str
+
+    def __setattr__(self, key, value):
+        """Converts a model object value to a gql node id"""
+        if isinstance(value, models.Model):
+            value = graphene.Node.to_global_id(value.__class__.__name__, value.pk)
+        elif isinstance(value, dict) and len(value) == 1:
+            _, value = value.popitem()
+        return super().__setattr__(key, value)
+
+
+@pytest.mark.parametrize(
+    "_test_name, event, expected_event_data",
+    [
+        (
+            "customer_placed_order_event",
```

Comment:
Not easy to make something simple and easy to maintain... I have an idea on how to make it KISS, I will do that tomorrow.

My only trouble is how to make sure that every event is getting tested. What do you think about something like this:
```python

@pytest.fixture(scope="module", autouse=True)
def are_all_events_are_tested():
    tested_events = set()
    yield lambda event: tested_events.add(event)
    assert tested_events == CustomerEvents.CHOICES.keys()

Reply:
I don't have any nice solution right now but I'm not sure if there is a need for such verification. It's rather a matter of code review to check if there are proper tests for each necessary scenario. This fixture looks very cool and I guess it could work but it's a bit complex concept and may be hard to understand/maintain. Anyways, these tests seemed a bit too complicated at a first glance, so I just wanted to suggest a simplification as they especially should be easy to read. But I have to say your ideas are really awesome! I haven't seen such usage of a fixture before

Reply:
I agree with @pawelzar - we don't want to introduce new patterns in code suddenly in one file, while all of the others use simple functions. If you're adding any new code and there are already some established patterns around that in the codebase, we have to follow them. Otherwise, we're just introducing inconsistency. Here we want to have just a simple unit functions for every case.

Some patterns might not work and we can think of improvements, but that has to be always discussed first and introduced consistently.